ucp/_libs/tests/test_rma.py

```diff
@@ -0,0 +1,140 @@
+    packed_rkey = mem.pack_rkey()
+    worker = ucx_api.UCXWorker(ctx)
+    ep = ucx_api.UCXEndpoint.create_from_worker_address(
+        worker, worker.get_address(), endpoint_error_handling=False
+    )
+    rkey = ep.unpack_rkey(packed_rkey)
+    self_mem = ucx_api.RemoteMemory(rkey, mem.address, msg_size)
+
+    send_msg = bytes(os.urandom(msg_size))
+    if not self_mem.put_nbi(send_msg):
+        blocking_flush(ep)
+    recv_msg = bytearray(len(send_msg))
+    if not self_mem.get_nbi(recv_msg):
+        blocking_flush(ep)
+    assert send_msg == recv_msg
+
+
+@pytest.mark.parametrize("msg_size", [10, 2 ** 24])
+def test_explicit(msg_size):
+    ctx = ucx_api.UCXContext({})
+    mem = ucx_api.UCXMemoryHandle.alloc(ctx, msg_size)
+    packed_rkey = mem.pack_rkey()
+    worker = ucx_api.UCXWorker(ctx)
+    ep = ucx_api.UCXEndpoint.create_from_worker_address(
+        worker, worker.get_address(), endpoint_error_handling=False
+    )
+    rkey = ep.unpack_rkey(packed_rkey)
+    self_mem = ucx_api.RemoteMemory(rkey, mem.address, msg_size)
+
+    send_msg = bytes(os.urandom(msg_size))
+    put_req = self_mem.put_nb(send_msg, _)
+    if put_req is not None:
+        blocking_flush(ep)
+    recv_msg = bytearray(len(send_msg))
+    recv_req = self_mem.get_nb(recv_msg, _)
+    if recv_req is not None:
+        blocking_flush(ep)
+    assert send_msg == recv_msg
+
+
+@pytest.mark.parametrize("msg_size", [10, 2 ** 24])
+def test_ucxio(msg_size):
+    ctx = ucx_api.UCXContext({})
+    mem = ucx_api.UCXMemoryHandle.alloc(ctx, msg_size)
+    packed_rkey = mem.pack_rkey()
+    worker = ucx_api.UCXWorker(ctx)
+    ep = ucx_api.UCXEndpoint.create_from_worker_address(
+        worker, worker.get_address(), endpoint_error_handling=False
+    )
+    rkey = ep.unpack_rkey(packed_rkey)
+
+    uio = ucx_api.UcxIO(ep, mem.address, msg_size, rkey)
+    send_msg = bytes(os.urandom(msg_size))
+    uio.write(send_msg)
+    uio.seek(0)
+    recv_msg = uio.read(msg_size)
+    assert send_msg == recv_msg
+    del uio
+
+
+def test_force_requests():
+    msg_size = 1024
+    ctx = ucx_api.UCXContext({})
+    mem = ucx_api.UCXMemoryHandle.alloc(ctx, msg_size)
+    packed_rkey = mem.pack_rkey()
+    worker = ucx_api.UCXWorker(ctx)
+    ep = ucx_api.UCXEndpoint.create_from_worker_address(
+        worker, worker.get_address(), endpoint_error_handling=False
+    )
+    rkey = ep.unpack_rkey(packed_rkey)
+    self_mem = ucx_api.RemoteMemory(rkey, mem.address, msg_size)
+
+    counter = 0
+    send_msg = bytes(os.urandom(msg_size))
+    req = self_mem.put_nb(send_msg, _)
+    while req is None:
+        counter = counter + 1
+        req = self_mem.put_nb(send_msg, _)
+        # This `if` is here because some combinations of transports, such as
+        # normal desktop PCs, will never have their transports exhausted. So
+        # we have a break to make sure this test still completes
+        if counter > 10000:
+            return
```

Comment:
I don't quite understand what's being done here. You first `put_nb` (outside of the loop), and then you count up to `10000` while `put_nb` is called every time again. Could you elaborate on what's this test really attempting to achieve? Furthermore, if you have to break here so that the test completes, what's the correct way to use this in systems such as desktop PCs as mentioned in the comment?

Reply:
The goal of this test is to hammer the put function until it runs out of resources and fails and goes into the request code path. On some transports like ugni and ib you will have a number of hardware resources where the op completes inline as a buffered send. The usual 'fire once and see if it works' style test will test this code path fine, but the code path that generates requests touches more code and is generally less tested. So this test will try and induce this behavior intentionally. 

The break is in there because on some transports like 'self' the test will never exhaust hardware resources as mentioned above because put will ultimately proxy into `memcpy()` and you will never get into a situation where you get a request, so the put will succeed forever and leave the test in an infinite loop.

Reply:
Sorry Matt for the delay, I'm on PTO so it's taking me longer to get to see PRs.

On cases such as `self` that you mentioned above, what will happen then if it will never exhaust, is it invalid to use `put_nb`? And if so, should we instead skip this test when certain transports are in use?

Reply:
That's fine, working on fixing the IO problem on ubuntu18. Which seems odd it's just on that one. But I can reproduce problems on my desktop, so it's a real thing.

Anyways, I should explain part of what I'm trying to do here at the low level driver/UCT layer. With a NIC like Cray Gemini/Aries or Mellanox IB the NIC will have a certain amount of hardware resources to send messages to remote NICs asynchronously without the host CPU getting involved. They are finite though, and if you do too many sends you can end up in a situation where UCP is using a queuing mechanism in UCT so that when a worker is progressed the worker can drive the TX queue, then try and empty the pending queue onto the TX queue. All of which involves callbacks. The goal of this test is to make sure that UCX gets into this state and then can be drained and everything that callbacks into Python works correctly.

The trouble with self is that since it doesn't have dedicated resources it has to use the host CPU and use it with `memcpy()`. It's still valid to call `put_nb` with self, it's just that it will never end up in the queuing code path. I don't want to try and restrict these mechanisms by particular transport, as that would run into problems with future transports and transports that aren't tested regularly. I'd rather have the test bang it's head into a wall and then decide, "If I didn't get to the queue code path by now, it is not going to happen ever."

Reply:
> I don't want to try and restrict these mechanisms by particular transport, as that would run into problems with future transports and transports that aren't tested regularly. I'd rather have the test bang it's head into a wall and then decide, "If I didn't get to the queue code path by now, it is not going to happen ever."

That's mostly fair, the only thing I don't like is this "silent success". Would it make more sense if instead of `return` we write `pytest.xfail` with some description of what's happening there?

Reply:
> That's mostly fair, the only thing I don't like is this "silent success". Would it make more sense if instead of return we write pytest.xfail with some description of what's happening there?
That sounds like the right way to go about that, though I have no idea how to use the `pytest.xfail` mechanism.